libraries/AP_GPS/AP_GPS.cpp

```diff
@@ -873,14 +873,48 @@ void AP_GPS::update(void)
 #endif
 }
 
+/*
+    determines if any gps's are degraded
+*/
+uint8_t AP_GPS::update_and_count_degraded_gps(void) {
+    uint8_t count = 0;
+    for (uint8_t i = 0; i < GPS_MAX_RECEIVERS; i++) {
+        if(state[i].status < GPS_OK_FIX_3D) {
+            _degraded_gps[i] = true;
+            count++;
+        }
+    }
+    return count;
+}
+
 /*
   update primary GPS instance
  */
 void AP_GPS::update_primary(void)
 {
 #if defined(GPS_BLENDED_INSTANCE)
+    uint8_t num_failed = update_and_count_degraded_gps();
+
+    if(num_failed > 0 && GPS_MAX_RECEIVERS > 1) {
+        // If the first GPS failed, use the second.
+        if(_degraded_gps[0]) {
+            _auto_switch = GPS_AUTO_SWITCH_USESECOND;
+        }
+
+        // If the second GPS failed, use the first.
+        if(_degraded_gps[1]) {
+            _auto_switch = GPS_AUTO_SWITCH_USEFIRST;
+        }
+    }
+
+    // All gps's have failed at some point, use the best moving forward
+    if(num_failed == GPS_MAX_RECEIVERS) {
+        _auto_switch = GPS_AUTO_SWITCH_USEBEST;
+    }
```

Comment:
Should this be evaluated first? With the current logic, if both GPSs have failed, it looks like we'll only check if GPS1 failed first, change `_auto_switch` to use second, then check if GPS2 also failed, which then changes `_auto_switch` to use first, then we check if both (all) GPSs have failed/degraded to < 3D fix and only then switch to use best.

Reply:
+1

Reply:
Sounds good